File: api/app/schemas/qualitative_factors.py
"""Pydantic schemas for Qualitative Risk Factor Configuration API."""
from datetime import datetime
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict, field_validator


# ============================================================================
# Guidance Schemas
# ============================================================================

class GuidanceCreate(BaseModel):
    """Schema for creating guidance."""
    rating: str = Field(..., pattern="^(HIGH|MEDIUM|LOW|VERY_HIGH|VERY_LOW)$")
    points: int = Field(..., ge=0, le=10)
    description: str = Field(..., min_length=1)
    sort_order: int = Field(default=0)


class GuidanceUpdate(BaseModel):
    """Schema for updating guidance."""
    rating: Optional[str] = Field(None, pattern="^(HIGH|MEDIUM|LOW|VERY_HIGH|VERY_LOW)$")
    points: Optional[int] = Field(None, ge=0, le=10)
    description: Optional[str] = None
    sort_order: Optional[int] = None


class GuidanceResponse(BaseModel):
    """Response schema for guidance."""
    model_config = ConfigDict(from_attributes=True)

    guidance_id: int
    factor_id: int
    rating: str
    points: int
    description: str
    sort_order: int


# ============================================================================
# Factor Schemas
# ============================================================================

class FactorCreate(BaseModel):
    """Schema for creating a factor."""
    code: str = Field(..., min_length=1, max_length=50)
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    weight: float = Field(..., ge=0.0, le=1.0)
    sort_order: int = Field(default=0)
    guidance: Optional[List[GuidanceCreate]] = None

    @field_validator('weight')
    @classmethod
    def validate_weight_precision(cls, v):
        """Ensure weight has reasonable precision."""
        return round(v, 4)


class FactorUpdate(BaseModel):
    """Schema for updating a factor."""
    code: Optional[str] = Field(None, min_length=1, max_length=50)
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    weight: Optional[float] = Field(None, ge=0.0, le=1.0)
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None

    @field_validator('weight')
    @classmethod
    def validate_weight_precision(cls, v):
        """Ensure weight has reasonable precision."""
        if v is not None:
            return round(v, 4)
        return v


class WeightUpdate(BaseModel):
    """Schema for updating weight only."""
    weight: float = Field(..., ge=0.0, le=1.0)

    @field_validator('weight')
    @classmethod
    def validate_weight_precision(cls, v):
        """Ensure weight has reasonable precision."""
        return round(v, 4)


class FactorResponse(BaseModel):
    """Response schema for a factor with guidance."""
    model_config = ConfigDict(from_attributes=True)

    factor_id: int
    code: str
    name: str
    description: Optional[str] = None
    weight: float
    sort_order: int
    is_active: bool
    guidance: List[GuidanceResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


# ============================================================================
# Utility Schemas
# ============================================================================

class WeightValidationResponse(BaseModel):
    """Response for weight validation."""
    valid: bool
    total: float
    message: Optional[str] = None


class ReorderRequest(BaseModel):
    """Request schema for reordering factors."""
    factor_ids: List[int]

File: web/src/api/qualitativeFactors.ts
/**
 * API client for Qualitative Risk Factor Configuration endpoints.
 */
import client from './client';

// ============================================================================
// Types
// ============================================================================

export interface GuidanceCreate {
    rating: 'HIGH' | 'MEDIUM' | 'LOW' | 'VERY_HIGH' | 'VERY_LOW';
    points: number;
    description: string;
    sort_order?: number;
}

export interface GuidanceUpdate {
    rating?: 'HIGH' | 'MEDIUM' | 'LOW' | 'VERY_HIGH' | 'VERY_LOW';
    points?: number;
    description?: string;
    sort_order?: number;
}

export interface GuidanceResponse {
    guidance_id: number;
    factor_id: number;
    rating: string;
    points: number;
    description: string;
    sort_order: number;
}

export interface FactorCreate {
    code: string;
    name: string;
    description?: string;
    weight: number;
    sort_order?: number;
    guidance?: GuidanceCreate[];
}

export interface FactorUpdate {
    code?: string;
    name?: string;
    description?: string;
    weight?: number;
    sort_order?: number;
    is_active?: boolean;
}

export interface WeightUpdate {
    weight: number;
}

export interface FactorResponse {
    factor_id: number;
    code: string;
    name: string;
    description: string | null;
    weight: number;
    sort_order: number;
    is_active: boolean;
    guidance: GuidanceResponse[];
    created_at: string;
    updated_at: string;
}

export interface WeightValidationResponse {
    valid: boolean;
    total: number;
    message: string | null;
}

export interface ReorderRequest {
    factor_ids: number[];
}

// ============================================================================
// API Functions
// ============================================================================

const BASE_URL = '/risk-assessment/factors';

/**
 * List all qualitative risk factors.
 */
export async function listFactors(includeInactive = false): Promise<FactorResponse[]> {
    const response = await client.get(`${BASE_URL}/`, {
        params: { include_inactive: includeInactive }
    });
    return response.data;
}

/**
 * Get a specific factor.
 */
export async function getFactor(factorId: number): Promise<FactorResponse> {
    const response = await client.get(`${BASE_URL}/${factorId}`);
    return response.data;
}

/**
 * Create a new factor.
 */
export async function createFactor(data: FactorCreate): Promise<FactorResponse> {
    const response = await client.post(`${BASE_URL}/`, data);
    return response.data;
}

/**
 * Update a factor.
 */
export async function updateFactor(factorId: number, data: FactorUpdate): Promise<FactorResponse> {
    const response = await client.put(`${BASE_URL}/${factorId}`, data);
    return response.data;
}

/**
 * Update only the weight of a factor.
 */
export async function updateFactorWeight(factorId: number, weight: number): Promise<FactorResponse> {
    const response = await client.patch(`${BASE_URL}/${factorId}/weight`, { weight });
    return response.data;
}

/**
 * Soft-delete a factor (set is_active=false).
 */
export async function deleteFactor(factorId: number): Promise<FactorResponse> {
    const response = await client.delete(`${BASE_URL}/${factorId}`);
    return response.data;
}

/**
 * Validate that active factor weights sum to 1.0.
 */
export async function validateWeights(): Promise<WeightValidationResponse> {
    const response = await client.post(`${BASE_URL}/validate-weights`);
    return response.data;
}

/**
 * Reorder factors by providing new order of IDs.
 */
export async function reorderFactors(factorIds: number[]): Promise<FactorResponse[]> {
    const response = await client.post(`${BASE_URL}/reorder`, { factor_ids: factorIds });
    return response.data;
}

/**
 * Add guidance to a factor.
 */
export async function addGuidance(factorId: number, data: GuidanceCreate): Promise<GuidanceResponse> {
    const response = await client.post(`${BASE_URL}/${factorId}/guidance`, data);
    return response.data;
}

/**
 * Update guidance.
 */
export async function updateGuidance(guidanceId: number, data: GuidanceUpdate): Promise<GuidanceResponse> {
    const response = await client.put(`${BASE_URL}/guidance/${guidanceId}`, data);
    return response.data;
}

/**
 * Delete guidance.
 */
export async function deleteGuidance(guidanceId: number): Promise<void> {
    await client.delete(`${BASE_URL}/guidance/${guidanceId}`);
}

export default {
    listFactors,
    getFactor,
    createFactor,
    updateFactor,
    updateFactorWeight,
    deleteFactor,
    validateWeights,
    reorderFactors,
    addGuidance,
    updateGuidance,
    deleteGuidance,
};
